panic: invalid memory address or nil pointer dereference in promhttp.HandlerForTransactional.func1

Hi!

the following code

```go
			registry := prometheus.NewRegistry()
			registry.MustRegister(collectorAdapter{*perfDataCollector})

			rw := httptest.NewRecorder()
			promhttp.HandlerFor(registry, promhttp.HandlerOpts{ErrorHandling: promhttp.ContinueOnError}).ServeHTTP(rw, &http.Request{})
			got := rw.Body.String()
```

ref: https://github.com/prometheus-community/windows_exporter/blob/a12486a2004876ed2e3a3c877c097602ee59710d/internal/collector/performancecounter/performancecounter_test_test.go#L202-L211

works fine before 1.24 and produces now an panic

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered, repanicked]
[signal 0xc0000005 code=0x0 addr=0x58 pc=0x7ff675fcd58e]

goroutine 35 [running]:
testing.tRunner.func1.2({0x7ff67647b500, 0x7ff676b02690})
	C:/hostedtoolcache/windows/go/1.26.5/x64/src/testing/testing.go:1974 +0x239
testing.tRunner.func1()
	C:/hostedtoolcache/windows/go/1.26.5/x64/src/testing/testing.go:1977 +0x349
panic({0x7ff67647b500?, 0x7ff676b02690?})
	C:/hostedtoolcache/windows/go/1.26.5/x64/src/runtime/panic.go:860 +0x13a
net/url.(*URL).Query(0x249fef2eae0?)
	C:/hostedtoolcache/windows/go/1.26.5/x64/src/net/url/url.go:1155 +0xe
github.com/prometheus/client_golang/prometheus/promhttp.HandlerForTransactional.func1({0x7ff6765b6b70, 0x1f07ab620980}, 0x1f07ab699400)
	C:/Users/runneradmin/go/pkg/mod/github.com/prometheus/client_golang@v1.24.0/prometheus/promhttp/http.go:377 +0xab2
net/http.HandlerFunc.ServeHTTP(0x0?, {0x7ff6765b6b70?, 0x1f07ab620980?}, 0x0?)
	C:/hostedtoolcache/windows/go/1.26.5/x64/src/net/http/server.go:2286 +0x29
github.com/prometheus-community/windows_exporter/internal/collector/performancecounter_test.TestCollector.func1(0x1f07ab742400)
	D:/a/windows_exporter/windows_exporter/internal/collector/performancecounter/performancecounter_test_test.go:206 +0x5fc
testing.tRunner(0x1f07ab742400, 0x1f07ab7282e0)
	C:/hostedtoolcache/windows/go/1.26.5/x64/src/testing/testing.go:2036 +0xc3
created by testing.(*T).Run in goroutine 34
	C:/hostedtoolcache/windows/go/1.26.5/x64/src/testing/testing.go:2101 +0x4a9
```

--- a/prometheus/promhttp/http.go
+++ b/prometheus/promhttp/http.go
@@ -372,12 +372,15 @@ func HandlerForTransactional(reg prometheus.TransactionalGatherer, opts HandlerO
 			return false
 		}
 
-		// Build metric name filter set from query params (if any)
+		// Build metric name filter set from query params (if any). The URL
+		// can be nil on hand-constructed requests.
 		var metricFilter map[string]struct{}
-		if metricNames := req.URL.Query()["name[]"]; len(metricNames) > 0 {
-			metricFilter = make(map[string]struct{}, len(metricNames))
-			for _, name := range metricNames {
-				metricFilter[name] = struct{}{}
+		if req.URL != nil {
+			if metricNames := req.URL.Query()["name[]"]; len(metricNames) > 0 {
+				metricFilter = make(map[string]struct{}, len(metricNames))
+				for _, name := range metricNames {
+					metricFilter[name] = struct{}{}
+				}
 			}
 		}
 

--- a/prometheus/promhttp/http_test.go
+++ b/prometheus/promhttp/http_test.go
@@ -879,6 +879,28 @@ func TestHandlerWithMetricFilter(t *testing.T) {
 	}
 }
 
+func TestHandlerWithNilRequestURL(t *testing.T) {
+	reg := prometheus.NewRegistry()
+
+	counter := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "test_counter",
+		Help: "A test counter.",
+	})
+	reg.MustRegister(counter)
+	counter.Inc()
+
+	writer := httptest.NewRecorder()
+	handler := HandlerFor(reg, HandlerOpts{ErrorHandling: ContinueOnError})
+	handler.ServeHTTP(writer, &http.Request{})
+
+	if got, want := writer.Code, http.StatusOK; got != want {
+		t.Errorf("got HTTP status code %d, want %d", got, want)
+	}
+	if body := writer.Body.String(); !strings.Contains(body, "test_counter") {
+		t.Errorf("expected body to contain test_counter, got: %s", body)
+	}
+}
+
 // syncGatherCounter is a thread-safe TransactionalGatherer wrapper that counts
 // Gather and done invocations. Safe for concurrent use from multiple goroutines,
 // unlike mockTransactionGatherer whose counters are not race-safe.

